example_allocators/bump_allocator/bump_alloc.c

```diff
@@ -51,16 +51,23 @@ void init_alloc(int size_in_bytes)
 char *bump_alloc(int len)
 {
 	char *chunk = 0;
+	size_t rounded_len; /* for CHERI alignment */
+
 	if (count + len <= max)
 	{
 		/* bump the pointer */
 		chunk = buffer + count;
 
+		/* check we can represent the capability accurately,
+		 * see p30 of CHERI C/C++ Prog Guide */
+		chunk = __builtin_align_up(chunk, ~cheri_representable_alignment_mask(len) + 1);
+		rounded_len = cheri_representable_length(len);
+
 		/* restrict capability range before returning ptr */
-		chunk = cheri_bounds_set_exact(chunk, len);
+		chunk = cheri_bounds_set_exact(chunk, rounded_len);
 
 		/* update bytes allocated count */
-		count += len;
+		count += rounded_len;
```

Comment:
Just an observation on usage (not a problem with the example): the caller should query the capability length, rather than assume that the buffer has exactly `len` bytes, for example if they want to zero the buffer or otherwise poison it. The [VIXL W^X example](https://git.linaro.org/arm/vixl.git/tree/examples/aarch64/morello-wx.cc?h=morello#n132) pads with known-undefined instructions because I think that's what a JIT would do as a security precaution.